lib/build.js

```diff
@@ -594,29 +586,62 @@ class BuildModel extends BaseModel {
     getMetrics(config = { stepName: null }) {
         // add build createTime to step metrics
         const createTime = this.createTime;
-        // filter out old steps without id and buildId info
-        let steps = this.steps;
 
-        // If stepName, get only that step
-        if (config.stepName) {
-            steps = steps.filter(s => s.name === config.stepName);
-        }
+        // filter out old steps without id and buildId info
+        return this.getSteps().then((steps) => {
+            let targetSteps = steps;
 
-        // Generate metrics
-        const metrics = steps.map((s) => {
-            // For older steps, add buildId info to it
-            if (!s.buildId) {
-                s.buildId = this.id;
+            // If stepName, get only that step
+            if (config.stepName) {
+                targetSteps = steps.filter(s => s.name === config.stepName);
             }
-            const { id, name, code, startTime, endTime } = s;
-            const duration = startTime && endTime
-                ? dayjs(endTime).diff(dayjs(startTime), 'second')
-                : null;
 
-            return { id, name, code, duration, createTime };
+            // Generate metrics
+            return Promise.all(targetSteps.map((s) => {
+                // For older steps, add buildId info to it
+                if (!s.buildId) {
+                    s.buildId = this.id;
+                }
+                const { id, name, code, startTime, endTime } = s;
+                const duration = startTime && endTime
+                    ? dayjs(endTime).diff(dayjs(startTime), 'second')
+                    : null;
+
+                return { id, name, code, duration, createTime };
+            }));
         });
+    }
 
-        return metrics;
+    /**
+     * Get a JSON representation of the build data with steps
+     * @method toJsonWithSteps
+     * @return {Promise} Resolves JSON of build with steps
+     */
+    toJsonWithSteps() {
+        // eslint-disable-next-line global-require
+        const StepFactory = require('./stepFactory');
+        const stepFactory = StepFactory.getInstance();
+
+        return stepFactory.list({
+            params: { buildId: this.id },
+            sortBy: 'id',
+            sort: 'ascending' })
+            .then((steps) => {
+                if (!steps.length) {
+                    throw new Error('Steps do not exist');
+                }
+
+                // This if statement should be removed after enough time has passed since build.steps removed.
+                // Make orders of steps in completed builds sure,
+                // because steps in old builds in DB have the order not sorted.
+                if (this.endTime) {
+                    steps.sort((a, b) =>
```

Comment:
`because steps in old builds in DB have the order not sorted.` can you elaborate this ? isn't `id` always ascending ? won't `sortBy` guarantee this ?

Reply:
In old builds, the order of records in steps table are not guaranteed. because it used `Promise.all` to save steps records, so each records is inserted asynchronously.
[fixed before](https://github.com/screwdriver-cd/models/blob/d380c78410630d57238d82c95f89a6d0f642c3f4/lib/buildFactory.js#L288)

After this fixed is merged, it guarantees the order.
[fixed after](https://github.com/sonic-screwdriver-cd/models/blob/8284c55cb627880130a81bc047c5d8262c6bbfdf/lib/buildFactory.js#L263)

So this is workaround for old builds.